### PR TITLE
expression: rewrite builtin function RAND, POW, SIGN, SQRT

### PR DESCRIPTION
**Pull request by breeswish**
to https://github.com/pingcap/tidb/issues/4080

Please refer to each commit for details.

### CODE REVIEW

File: expression/builtin_math.go [hunk 1]
```diff
@@ -773,38 +773,65 @@ type randFunctionClass struct {
...
 
 func (c *randFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	err := errors.Trace(c.verifyArgs(args))
-	bt := &builtinRandSig{baseBuiltinFunc: newBaseBuiltinFunc(args, ctx)}
-	bt.deterministic = false
-	return bt.setSelf(bt), errors.Trace(err)
+	if err := c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+	var sig builtinFunc
```
Comment (XuHuaiyu) [inline expression/builtin_math.go]: put this definition to the return value of this function.

File: expression/builtin_math.go [hunk 2]
```diff
@@ -773,38 +773,65 @@ type randFunctionClass struct {
...
+	if len(args) == 0 {
```
Comment (XuHuaiyu) [inline expression/builtin_math.go]: I'm not sure will the following code be better?
``` go
if len(args) == 0 {
    bf, err = 
}else{
    bf, err = 
}
if err != nil {
}
bt := 
bt.determinstic = false
if len(args) == 0{
   sig =  
}else{
    sig = 
}
return sig.setSelf(sig), nil
```

File: expression/builtin_math.go [hunk 3]
```diff
@@ -773,38 +773,65 @@ type randFunctionClass struct {
...
-		if len(args) == 1 && !args[0].IsNull() {
-			seed, err := args[0].ToInt64(b.ctx.GetSessionVars().StmtCtx)
-			if err != nil {
-				return d, errors.Trace(err)
-			}
-			b.randGen = rand.New(rand.NewSource(seed))
-		} else {
-			// If seed is not set, we use current timestamp as seed.
+		if isNull {
+			// When seed is NULL, it is equal to RAND()
```
Comment (XuHuaiyu) [inline expression/builtin_math.go]: add `.` at the end of this comment.

File: expression/builtin_math_test.go [hunk 4]
```diff
@@ -597,30 +596,36 @@ func (s *testEvaluatorSuite) TestConv(c *C) {
 func (s *testEvaluatorSuite) TestSign(c *C) {
```
Comment (XuHuaiyu) [inline expression/builtin_math_test.go]: add test for isDeterministic.
so as func Sqrt

File: expression/builtin_math.go (hunk 2 above)
Comment (breeswish) [inline expression/builtin_math.go]: I think the only advantage is that, the lines of code SEEMS to be shorter : )

**Comment by XuHuaiyu** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Pull request merged**